codalab/worker/bundle_manager.py

```diff
@@ -284,56 +285,56 @@ def _schedule_run_bundles_on_workers(self, workers, user_owned):
             else:
                 workers_list = workers.user_owned_workers(self._model.root_user_id)
 
-            workers_list = self._filter_and_sort_workers(workers_list, bundle)
+            # save _compute_request_X values because somehow they get mutated somewhere down the line (wtf!)
+            # i.e. multiple calls may return different values; which screws up scheduling really badly
+            # TODO: hunt down this bug
+            request_cpus = self._compute_request_cpus(bundle)
+            request_gpus = self._compute_request_gpus(bundle)
+            request_memory = self._compute_request_memory(bundle)
+
+            workers_list = self._filter_and_sort_workers(workers_list, bundle,
+                    request_cpus, request_gpus, request_memory)
 
             for worker in workers_list:
-                if self._try_start_bundle(workers, worker, bundle):
+                if self._try_start_bundle(workers, worker, bundle, request_cpus, request_gpus):
                     break
                 else:
                     continue  # Try the next worker.
 
-    def _deduct_worker_resources(self, workers_list):
-        """
-        From each worker, subtract resources used by running bundles. Modifies the list.
-        """
-        for worker in workers_list:
-            for uuid in worker['run_uuids']:
-                bundle = self._model.get_bundle(uuid)
-                worker['cpus'] -= self._compute_request_cpus(bundle) or 1
-                worker['gpus'] -= self._compute_request_gpus(bundle) or 0
-                #TODO: memory_bytes
-
-    def _filter_and_sort_workers(self, workers_list, bundle):
+    def _filter_and_sort_workers(self, workers_list, bundle, request_cpus, request_gpus, request_memory):
         """
         Filters the workers to those that can run the given bundle and returns
         the list sorted in order of preference for running the bundle.
-
-        Note: the worker slots field is no longer in use so it is effectively ignored
         """
 
-        # keep track of which workers have GPUs
-        has_gpu = {}
+        worker_has_gpu = {} # keep track of which workers have GPUs
+        worker_free_cpus = {}
+        worker_free_gpus = {}
+
+        # From each worker, initialize with the cpuset/gpuset lengths
         for worker in workers_list:
             worker_id = worker['worker_id']
-            has_gpu[worker_id] = True if worker['gpus'] > 0 else False
+            worker_has_gpu[worker_id] = len(worker['gpuset']) > 0
```

Comment:
why do we need this this? you can just compute from worker_free_gpus locally to avoid having as much state.

Reply:
We differentiate between workers having 0 gpus and workers having 0 _free_ gpus; this is so that we can schedule CPU-only bundles to CPU-only workers (aka 0 gpu workers) preferentially.